README
NAME
    ExtUtils::CppGuess - guess C++ compiler and flags

SYNOPSIS
    With Extutils::MakeMaker:

        use ExtUtils::CppGuess;
    
        my $guess = ExtUtils::CppGuess->new;
    
        WriteMakefile
          ( # MakeMaker args,
            $guess->makemaker_options,
            );

    With Module::Build:

        my $guess = ExtUtils::CppGuess->new;
    
        my $build = Module::Build->new
          ( # Module::Build arguments
            $guess->module_build_options,
            );
        $build->create_build_script;

DESCRIPTION
    "ExtUtils::CppGuess" attempts to guess the system's C++ compiler that is
    compatible with the C compiler that your perl was built with.

    It can generate the necessary options to the Module::Build constructor
    or to ExtUtils::MakeMaker's "WriteMakefile" function.

METHODS
  new
    Creates a new "ExtUtils::CppGuess" object. Takes the path to the C
    compiler as the "cc" argument, but falls back to the value of
    $Config{cc}, which should be what you want anyway.

    You can specify "extra_compiler_flags" and "extra_linker_flags" (as
    strings) which will be merged in with the auto-detected ones.

  module_build_options
    Returns the correct options to the constructor of "Module::Build". These
    are:

        extra_compiler_flags
        extra_linker_flags

  makemaker_options
    Returns the correct options to the "WriteMakefile" function of
    "ExtUtils::MakeMaker". These are:

        CCFLAGS
        dynamic_lib => { OTHERLDFLAGS => ... }

    If you specify the extra compiler or linker flags in the constructor,
    they'll be merged into "CCFLAGS" or "OTHERLDFLAGS" respectively.

  is_gcc
    Returns true if the detected compiler is in the gcc family.

  is_msvc
    Returns true if the detected compiler is in the MS VC family.

  add_extra_compiler_flags
    Takes a string as argument that is added to the string of extra compiler
    flags.

  add_extra_linker_flags
    Takes a string as argument that is added to the string of extra linker
    flags.

AUTHOR
    Mattia Barbon <mbarbon@cpan.org>

    Steffen Mueller <smueller@cpan.org>

LICENSE
    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
- fix test-failure caused if CXX is set - thanks @hroncok for report

0.25 2022-04-20
- ignore environment variables CXXFLAGS, CXXLDFLAGS (#26) - thanks @hroncok for report

0.24 2022-04-19
- support environment variables CXX, CXXFLAGS, CXXLDFLAGS - thanks @dpryan79 for suggestion

0.23 2021-05-31
- added CI - thanks @zmughal
- now recognises MacOS clang - thanks @xenu

0.22 2021-05-29
- Add C++ standard support flag detection - thanks @zmughal!
  See <https://github.com/tsee/extutils-cppguess/pull/24>.

0.21 2020-01-23
- no give $Config{ccflags} in Module::Build as add not replace - thanks @xenu

0.20    Sat 31 Aug 22:20:32 BST 2019
    - no link -lstdc++ if -static-libstdc++ - thanks @andygrundman

0.19    Mon  1 Apr 07:08:46 BST 2019
    - support clang on Perl 5.14-16 with ExtUtils::ParseXS >= 3.04_01 (yes, really)

0.18    Sun 31 Mar 17:49:21 BST 2019
    - improve clang detection

0.17    Sat 30 Mar 16:03:52 GMT 2019
    - support Sun Studio compiler family

0.16    Sat 30 Mar 00:09:42 GMT 2019
    - test output more if no clue about compiler

0.15    Fri 29 Mar 22:42:00 GMT 2019
    - iostream_fname, cpp_flavor_defs methods
    - add which actual compiler to use to generated MB and EUMM flags
    - make the test sensitive to whether to #include with a ".h"

0.14	Mon 25 Mar 21:13:36 GMT 2019
    - add is_clang, unify OS-based detection

0.13	Mon Mar 25 02:57:02 2019 +0000
    - repo tidied slightly
    - use Travis
    - add compiler_command, linker_flags
    - test all known configurations
    - inhale knowledge from Inline::CPP

0.12	Sat Oct 21 18:12:51 BST 2017
    - Fix tests on 5.26 without '.' in @INC (kentfredric)

0.11   Thu Sep 10 13:12:17 MDT 2015
    - Add leading whitespace to $cflags in _get_cflags (bulk88)

0.10   Tue Sep 08 21:15:30 MDT 2015
    - Config's ccflags must always be used. (bulk88)

0.09   Sat Apr 11 16:05:50 BST 2015
	- Convert to EUMM
	- Make tests divulge guessed options for better debugging


0.08   Thu Jan 15 03:07:38 GMT 2015
	- Update metadata


0.07   Thu Jul  7 17:00:00 CEST 2011
	- Fix for unknown enum `vtype' on solaris 2.11 (Tobias Leich)
          See http://www.cpantesters.org/cpan/report/a0cca9ee-39e3-11e0-83a5-34754e7aadc9
        - NetBSD fix for: Undefined PLT symbol _Unwind_Resume (Tobias Leich)


0.06   Sun Jun  5 21:00:00 CEST 2011
	- Fix for -D_FILE_OFFSET_BITS (Tobias Leich)
          See http://www.cpantesters.org/cpan/report/b6e29992-894c-11e0-b6b1-f9d75287cfae

0.05   Tue Feb 15 23:00:00 CEST 2010
	- Fix test failures on win32 (Shmuel Fomberg)

0.04   Mon Jul 12 18:00:00 CEST 2010
	- Fix problem with duplicate extra-linker/compiler options

0.03   Thu Jul  1 18:00:00 CEST 2010
	- Fix tests to use the current perl
        - Make tests use Capture::Tiny
        - Very explicit dependencies in Build.PL

0.02   Mon Jun 28 20:00:00 CEST 2010
	- Just removed debugging code.

0.01   Mon Jun 28 20:00:00 CEST 2010
	- First release.


t/module/Build.PL
#!/usr/bin/perl -w

use strict;
use blib '../..';
use Module::Build;
use ExtUtils::CppGuess;
use ExtUtils::ParseXS;

my $guess = ExtUtils::CppGuess->new;
$guess->add_extra_compiler_flags(
  '-DINCLUDE_DOT=' .
  ($guess->iostream_fname =~ /\./ ? 1 : 0)
);

# broken on 5.14, on EUPXS 3.04_01 to 3.35
if ($ExtUtils::ParseXS::VERSION > 3.04) {
  if ($] lt '5.016000') {
    $guess->add_extra_compiler_flags('-DCLANG_WORKAROUND_514');
  }
}

my $build = Module::Build->new
  ( module_name     => 'CppGuessTest',
    dist_abstract   => 'a test module',
    license         => 'perl',
    xs_files        => { 'CppGuessTest.xs' => 'lib/CppGuessTest.xs' },
    $guess->module_build_options,
    );
$build->create_build_script;


t/lib/TestUtils.pm
package t::lib::TestUtils;

use strict;

use ExtUtils::Manifest qw(manicopy maniread);
use File::Path qw(rmtree);
use File::Spec::Functions qw(rel2abs);
use Cwd qw(cwd);
use Fatal qw(chdir);
use Config qw();

require Exporter; *import = \&Exporter::import;

our @EXPORT = qw(prepare_test build_makemaker build_module_build);

sub prepare_test {
    my( $source, $destination ) = @_;
    my $cwd = cwd();

    $destination = rel2abs( $destination );
    rmtree( $destination );
    chdir( $source );
    manicopy( maniread(), $destination );
    chdir( $cwd );
}

sub build_module_build {
    my( $path ) = @_;
    my $cwd = cwd();
    chdir $path;

    my $build_pl = `$^X Build.PL 2>&1`;
    my $build_pl_ok = $?;

    if( $build_pl_ok != 0 ) {
        chdir( $cwd );
        return ( 0, $build_pl, undef, undef );
    }

    my $build = `$^X Build 2>&1`;
    my $build_ok = $?;

    if( $build_ok != 0 ) {
        chdir( $cwd );
        return ( 0, $build_pl, $build, undef );
    }

    my $build_test = `$^X Build test 2>&1`;
    my $build_test_ok = $?;

    if( $build_test_ok != 0 ) {
        chdir( $cwd );
        return ( 0, $build_pl, $build, $build_test );
    } else {
        chdir( $cwd );
        return ( 1, $build_pl, $build, $build_test );
    }
}

sub build_makemaker {
    my( $path ) = @_;
    my $cwd = cwd();
    chdir $path;

    my $makefile_pl = `$^X Makefile.PL 2>&1`;
    my $makefile_pl_ok = $?;

    if( $makefile_pl_ok != 0 ) {
        chdir( $cwd );
        return ( 0, $makefile_pl, undef, undef );
    }

    my $make = `$Config::Config{make} 2>&1`;
    my $make_ok = $?;

    if( $make_ok != 0 ) {
        chdir( $cwd );
        return ( 0, $makefile_pl, $make, undef );
    }

    my $make_test = `$Config::Config{make} test 2>&1`;
    my $make_test_ok = $?;

    if( $make_test_ok != 0 ) {
        chdir( $cwd );
        return ( 0, $makefile_pl, $make, $make_test );
    } else {
        chdir( $cwd );
        return ( 1, $makefile_pl, $make, $make_test );
    }
}

1;


t/module/Makefile.PL
#!/usr/bin/perl -w

use strict;
use blib '../..';
use ExtUtils::MakeMaker;
use ExtUtils::CppGuess;
use ExtUtils::ParseXS;

my $guess = ExtUtils::CppGuess->new;

# broken on 5.14, on EUPXS 3.04_01 to 3.35
if ($ExtUtils::ParseXS::VERSION > 3.04) {
  if ($] lt '5.016000') {
    $guess->add_extra_compiler_flags('-DCLANG_WORKAROUND_514');
  }
}

$guess->add_extra_compiler_flags(
  '-DINCLUDE_DOT=' .
  ($guess->iostream_fname =~ /\./ ? 1 : 0)
);

WriteMakefile
  ( NAME            => 'CppGuessTest',
    VERSION_FROM    => 'lib/CppGuessTest.pm',
    PL_FILES        => {},
    $guess->makemaker_options,
    );


Build.PL
#!/usr/bin/perl -w

use strict;
use Module::Build;

my $build = Module::Build->new(
    module_name    => 'ExtUtils::CppGuess',
    license        => 'perl',
    configure_requires => { 'Module::Build' => 0.42 },
    requires       => {
        'Capture::Tiny' => '0',
        'File::Basename' => '0',
    },
    build_requires  => {
        'Module::Build' => '0',
    },
    test_requires  => {
        'Capture::Tiny' => '0',
        'Module::Build' => '0',
        'ExtUtils::MakeMaker' => '0',
        'Test::More' => '0',
        'File::Spec' => '0',
        'ExtUtils::Manifest' => '0',
        'Fatal' => '0',
        'Cwd' => '0',
    },
    add_to_cleanup => [ 't/module_build', 't/makemaker' ],
    meta_merge => {
        resources => {
            repository => 'git://github.com/tsee/extutils-cppguess'
        }
    },
);

$build->create_build_script;


Makefile.PL
#!/usr/bin/perl -w

use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME => 'ExtUtils::CppGuess',
    VERSION_FROM => 'lib/ExtUtils/CppGuess.pm',
    LICENSE => 'perl',
    PREREQ_PM => {
        'Capture::Tiny' => '0',
        'File::Basename' => '0',
        'File::Spec' => '0',
        'File::Temp' => '0',
    },
    TEST_REQUIRES  => {
        'Capture::Tiny' => '0',
        'Module::Build' => '0',
        'ExtUtils::CBuilder' => '0.280231',
        'ExtUtils::MakeMaker' => '0',
        'Test::More' => '0.88',
        'ExtUtils::Manifest' => '0',
        'Fatal' => '0',
        'Cwd' => '0',
        'Data::Dumper' => 0,
    },
    clean => { FILES => 't/module_build t/makemaker' },
    META_MERGE => {
      "meta-spec" => { version => 2 },
      resources => {
        bugtracker => { web => 'https://github.com/tsee/extutils-cppguess/issues' },
        repository => {
          type => 'git',
          url => 'git://github.com/tsee/extutils-cppguess',
          web => 'https://github.com/tsee/extutils-cppguess',
        },
      },
    },
);


lib/ExtUtils/CppGuess.pm
package ExtUtils::CppGuess;

use strict;
use warnings;

=head1 NAME

ExtUtils::CppGuess - guess C++ compiler and flags

=head1 SYNOPSIS

With L<Extutils::MakeMaker>:

    use ExtUtils::CppGuess;

    my $guess = ExtUtils::CppGuess->new;

    WriteMakefile
      ( # MakeMaker args,
        $guess->makemaker_options,
        );

With L<Module::Build>:

    my $guess = ExtUtils::CppGuess->new;

    my $build = Module::Build->new
      ( # Module::Build arguments
        $guess->module_build_options,
        );
    $build->create_build_script;

=head1 DESCRIPTION

C<ExtUtils::CppGuess> attempts to guess the system's C++ compiler
that is compatible with the C compiler that your perl was built with.

It can generate the necessary options to the L<Module::Build>
constructor or to L<ExtUtils::MakeMaker>'s C<WriteMakefile>
function.

=head1 ENVIRONMENT

As of 0.24, the environment variable C<CXX>
defines the obvious value, and will be used instead of any detection.
Supplied arguments to L</new> will still win.

=head1 METHODS

=head2 new

Creates a new C<ExtUtils::CppGuess> object.
Takes the path to the C compiler as the C<cc> argument,
but falls back to the value of C<$Config{cc}>, which should
be what you want anyway.

You can specify C<extra_compiler_flags> and C<extra_linker_flags>
(as strings) which will be merged in with the auto-detected ones.

=head2 module_build_options

Returns the correct options to the constructor of C<Module::Build>.
These are:

    extra_compiler_flags
    extra_linker_flags
    config => { cc => ... }, # as of 0.15

Please note the above may have problems on Perl <= 5.8 with
L<ExtUtils::CBuilder> <= 0.280230 due to a Perl RE issue.

=head2 makemaker_options

Returns the correct options to the C<WriteMakefile> function of
C<ExtUtils::MakeMaker>.
These are:

    CCFLAGS
    dynamic_lib => { OTHERLDFLAGS => ... }
    CC # as of 0.15

If you specify the extra compiler or linker flags in the
constructor, they'll be merged into C<CCFLAGS> or
C<OTHERLDFLAGS> respectively.

=head2 is_gcc

Returns true if the detected compiler is in the gcc family.

=head2 is_msvc

Returns true if the detected compiler is in the MS VC family.

=head2 is_clang

Returns true if the detected compiler is in the Clang family.

=head2 is_sunstudio

Returns true if the detected compiler is in the Sun Studio family.

=head2 add_extra_compiler_flags

Takes a string as argument that is added to the string of extra compiler
flags.

=head2 add_extra_linker_flags

Takes a string as argument that is added to the string of extra linker
flags.

=head2 compiler_command

Returns the string that can be passed to C<system> to execute the compiler.
Will include the flags returned as the Module::Build
C<extra_compiler_flags>.

Added in 0.13.

=head2 linker_flags

The same as returned as the Module::Build C<extra_linker_flags>.

Added in 0.13.

=head2 iostream_fname

Returns the filename to C<#include> to get iostream capability.

This can be used a bit creatively to be portable in one's XS files,
as the tests for this module need to be:

  # in Makefile.PL:
  $guess->add_extra_compiler_flags(
    '-DINCLUDE_DOT=' .
    ($guess->iostream_fname =~ /\./ ? 1 : 0)
  );

  // in your .xs file:
  #if INCLUDE_DOT
  #include <string.h>
  #else
  #include <string>
  #endif

Added in 0.15.

=head2 cpp_flavor_defs

Returns the text for a header that C<#define>s
C<__INLINE_CPP_STANDARD_HEADERS> and C<__INLINE_CPP_NAMESPACE_STD> if
the standard headers and namespace are available. This is determined by
trying to compile C++ with C<< #define <iostream> >> - if it succeeds,
the symbols will be defined, else commented.

Added in 0.15.

=head2 cpp_standard_flag

  $guess->cpp_standard_flag( $standard_name )

Given a string C<$standard_name> that is currently one of

=over

=item * C<< C++98 >>

=item * C<< C++11 >>

=item * C<< C++14 >>

=item * C<< C++17 >>

=back

returns a string with a flag that can be used to tell the compiler to support
that version of the C++ standard or dies if version is not supported.

Added in version v0.22.

=head1 AUTHOR

Mattia Barbon <mbarbon@cpan.org>

Steffen Mueller <smueller@cpan.org>

Tobias Leich <froggs@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright 2010, 2011 by Mattia Barbon.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut

use Config ();
use File::Basename qw();
use Capture::Tiny 'capture_merged';
use File::Spec::Functions qw(catfile);
use File::Temp qw(tempdir);

our $VERSION = '0.24';

sub new {
    my( $class, %args ) = @_;
    my $self = bless \%args, $class;

    # Allow override of default %Config::Config; useful in testing.
    if( !defined $self->{config} ) {
      if ($ExtUtils::MakeMaker::Config::VERSION) {
        # tricksy hobbitses are overriding Config, go with it
        $self->{config} = \%ExtUtils::MakeMaker::Config::Config;
      } else {
        $self->{config} = \%Config::Config;
      }
    }

    for (['cc','cc',$Config::Config{cc}], ['os','osname',$^O], ['osvers','osvers','']) {
      my ($key, $confkey, $fallback) = @$_;
      next if defined $self->{$key};
      $self->{$key} =
        defined $self->{config}{$confkey} ? $self->{config}{$confkey} : $fallback;
    }
    return $self;
}

# Thus saith the law: All references to %Config::Config shall come through
# $self->_config.  Accessors shall provide access to key components thereof.
# Testing shall thus grow stronger, verifying performance for platforms diverse
# to which access we have not.

sub _config { shift->{config} }
sub _cc     { shift->{cc}     }
sub _os     { shift->{os}     }
sub _osvers { shift->{osvers} }

our %ENV2VAL = (
  CXX => 'compiler_command',
);
# This is IBM's "how to compile on" list with lots of compilers:
# https://www.ibm.com/support/knowledgecenter/en/SS4PJT_5.2.0/com.ibm.help.cd52.unix.doc/com.ibm.help.cdunix_user.doc/CDU_Compiling_Custom_Programs.html
sub guess_compiler {
  my $self = shift;
  return $self->{guess} if $self->{guess};
  my $c_compiler = $self->_cc;
#  $c_compiler = $Config::Config{cc} if not defined $c_compiler;
  my %guess;
  if ($self->{os} eq 'freebsd' && $self->{osvers} =~ /^(\d+)/ && $1 >= 10) {
    $self->{is_clang} = 1; # special-case override
    %guess = (
      compiler_command => 'clang++',
      extra_lflags => '-lc++',
    );
  } elsif( $self->_cc_is_sunstudio( $c_compiler ) ) {
    %guess = (
      compiler_command => 'CC',
      extra_cflags => '',
      extra_lflags => '',
    );
  } elsif( $self->_cc_is_clang( $c_compiler ) ) {
    %guess = (
      compiler_command => 'clang++',
      extra_cflags => '-xc++',
      extra_lflags => '-lstdc++',
    );
  } elsif( $self->_cc_is_gcc( $c_compiler ) ) {
    %guess = (
      compiler_command => 'g++',
      extra_cflags => '-xc++',
    );
    # Don't use -lstdc++ if Perl was linked with -static-libstdc++ (ActivePerl 5.18+ on Windows)
    $guess{extra_lflags} = '-lstdc++'
      unless ($self->_config->{ldflags} || '') =~ /static-libstdc\+\+/;
  } elsif ( $self->_cc_is_msvc( $c_compiler ) ) {
    %guess = (
      compiler_command => 'cl',
      extra_cflags => '-TP -EHsc',
      extra_lflags => 'msvcprt.lib',
    );
  }
  $guess{$ENV2VAL{$_}} = $ENV{$_} for grep defined $ENV{$_}, keys %ENV2VAL;
  if (!%guess) {
    my $v1 = `$c_compiler -v`;
    my $v2 = `$c_compiler -V`;
    my $v3 = `$c_compiler --version`;
    my $os = $self->_os;
    die <<EOF;
Unable to determine a C++ compiler for '$c_compiler' on $os
Version attempts:
-v: '$v1'
-V: '$v2'
--version: '$v3'
EOF
  }
  $guess{extra_lflags} .= ' -lgcc_s'
    if $self->_os eq 'netbsd' and
    $guess{compiler_command} =~ /g\+\+/i and
    $guess{extra_lflags} !~ /-lgcc_s/;
  $self->{guess} = \%guess;
}

sub _get_cflags {
  my ($self, $omit_ccflags) = @_;
  $self->guess_compiler or die;
  join ' ', '', map _trim_whitespace($_), grep defined && length,
    ($omit_ccflags ? '' : $self->_config->{ccflags}),
    $self->{guess}{extra_cflags},
    $self->{extra_compiler_flags},
    ($self->is_clang ? '-Wno-reserved-user-defined-literal' : ()),
    ;
}

sub _get_lflags {
  my $self = shift;
  $self->guess_compiler || die;
  join ' ', '', map _trim_whitespace($_), grep defined && length,
    $self->{guess}{extra_lflags},
    $self->{extra_linker_flags},
    ;
}

sub makemaker_options {
    my $self = shift;

    my $lflags = $self->_get_lflags;
    my $cflags = $self->_get_cflags;

    return (
      CCFLAGS      => $cflags,
      dynamic_lib  => { OTHERLDFLAGS => $lflags },
      CC => $self->{guess}{compiler_command},
    );
}


sub module_build_options {
    my $self = shift;

    my $lflags = $self->_get_lflags;
    # We're omitting ccflags to avoid duplication of flags, because unlike
    # makemaker, we're appending to the compiler flags, not overriding
    # them. They already contain $Config{ccflags}.
    my $cflags = $self->_get_cflags(1);

    return (
      extra_compiler_flags => $cflags,
      extra_linker_flags   => $lflags,
      config => { cc => $self->{guess}{compiler_command} },
    );
}

# originally from Alien::wxWidgets::Utility
# Why was this hanging around outside of all functions, and without any other
# use of $quotes?
# my $quotes = $self->_os =~ /MSWin32/ ? '"' : "'";

sub _capture {
    my @cmd = @_;
    my $out = capture_merged { system(@cmd) };
    $out = '' if not defined $out;
    return $out;
}

# capture the output of a command that is run with piping
# to stdin of the command. We immediately close the pipe.
sub _capture_empty_stdin {
    my $cmd = shift;
    my $out = capture_merged {
        if ( open my $fh, '|-', $cmd ) {
          close $fh;
        }
    };
    $out = '' if not defined $out;
    return $out;
}


sub _cc_is_msvc {
    my( $self, $cc ) = @_;
    $self->{is_msvc}
      = ($self->_os =~ /MSWin32/ and File::Basename::basename($cc) =~ /^cl/i);
    return $self->{is_msvc};
}

sub _cc_is_gcc {
    my( $self, $cc ) = @_;
    $self->{is_gcc} = 0;
    my $cc_version = _capture( "$cc --version" );
    if (
         $cc_version =~ m/\bg(?:cc|\+\+)/i # 3.x, some 4.x
      || scalar( _capture( "$cc" ) =~ m/\bgcc\b/i ) # 2.95
      || scalar(_capture_empty_stdin("$cc -dM -E -") =~ /__GNUC__/) # more or less universal?
      || scalar($cc_version =~ m/\bcc\b.*Free Software Foundation/si) # some 4.x?
      || $cc eq 'gcc' # because why would they lie?
    ) {
      $self->{is_gcc} = 1;
    }
    return $self->{is_gcc};
}

sub _cc_is_clang {
    my( $self, $cc ) = @_;
    $self->{is_clang} = 0;
    my $cc_version = _capture( "$cc --version" );
    if (
         $cc_version =~ m/\A(?:clang|apple llvm)/i
      || $cc eq 'clang' # because why would they lie?
      || (($self->_config->{gccversion} || '') =~ /Clang|Apple LLVM/),
    ) {
      $self->{is_clang} = 1;
    }
    return $self->{is_clang};
}

sub _cc_is_sunstudio {
    my( $self, $cc ) = @_;
    $self->{is_sunstudio} = 0;
    my $cc_version = _capture( "$cc -V" );
    if (
         $cc_version =~ m/Sun C/i
      || $cc =~ /SUNWspro/ # because why would they lie?
    ) {
      $self->{is_sunstudio} = 1;
    }
    return $self->{is_sunstudio};
}

sub is_gcc {
    my $self = shift;
    $self->guess_compiler || die;
    return $self->{is_gcc};
}

sub is_msvc {
    my $self = shift;
    $self->guess_compiler || die;
    return $self->{is_msvc};
}

sub is_clang {
    my $self = shift;
    $self->guess_compiler || die;
    return $self->{is_clang};
}

sub is_sunstudio {
    my $self = shift;
    $self->guess_compiler || die;
    return $self->{is_sunstudio};
}

sub add_extra_compiler_flags {
    my( $self, $string ) = @_;

    $self->{extra_compiler_flags}
      = join ' ', map _trim_whitespace($_), grep defined && length,
        $self->{extra_compiler_flags}, $string;
}

sub add_extra_linker_flags {
    my( $self, $string ) = @_;
    $self->{extra_linker_flags}
      = join ' ', map _trim_whitespace($_), grep defined && length,
        $self->{extra_linker_flags}, $string;
}

sub compiler_command {
    my( $self ) = @_;
    $self->guess_compiler || die;
    my $cc = $self->{guess}{compiler_command};
    my $cflags = $self->_get_cflags;
    join ' ', map _trim_whitespace($_), grep defined && length, $cc, $cflags;
}

sub _trim_whitespace {
  my $string = shift;
  $string =~ s/^\s+|\s+$//g;
  return $string;
}

sub linker_flags {
    my( $self ) = @_;
    _trim_whitespace($self->_get_lflags);
}

sub _to_file {
  my ($file, @data) = @_;
  open my $fh, '>', $file
    or die "open $file: $!\n";
  print $fh @data or die "write $file: $!\n";
  close $fh or die "close $file: $!\n";
}

my $test_cpp_filename = 'ilcpptest';        # '.cpp' appended via open.
my $test_cpp          = <<'END_TEST_CPP';
#include <iostream>
int main(){ return 0; }
END_TEST_CPP

# Compile the given code and returns true on success.
#
# Can optionally be given compiler flags.
sub _can_compile_code {
  my( $self, $cpp_code, $compiler_flags ) = @_;
  my $dir = tempdir( CLEANUP => 1 );
  my $file = catfile( $dir, qq{$test_cpp_filename.cpp} );
  my $exe = catfile( $dir, qq{$test_cpp_filename.exe} );
  _to_file $file, $cpp_code;
  my $command = join ' ',
    $self->compiler_command,
    @{ defined $compiler_flags ? $compiler_flags : [] },
    ($self->is_msvc ? qq{-Fe:} : qq{-o }) . $exe,
    $file,
    ;
  return 0 == system $command;
}

# returns true if compile succeeded, false if failed
sub _compile_no_h {
  my( $self ) = @_;
  return $self->{no_h_status} if defined $self->{no_h_status};
  $self->guess_compiler || die;
  $self->{no_h_status} = $self->_can_compile_code( $test_cpp );
}

sub iostream_fname {
  my( $self ) = @_;
  'iostream' . ($self->_compile_no_h ? '' : '.h');
}

sub cpp_flavor_defs {
  my( $self ) = @_;
  my $comment = ($self->_compile_no_h ? '' : '//');
  sprintf <<'END_FLAVOR_DEFINITIONS', $comment, $comment;

%s#define __INLINE_CPP_STANDARD_HEADERS 1
%s#define __INLINE_CPP_NAMESPACE_STD 1

END_FLAVOR_DEFINITIONS
}

# Listed in order by year.
our @CPP_STANDARDS = (
  'C++98',
  'C++11',
  'C++14',
  'C++17',
);
# Hash of flags for each compiler:
#
# Structure
#  Hash:
#   - key: <detected compiler name string>
#   - value:
#       Hash:
#         - key: <C++ standard name string>
#         - value:
#             ArrayRef[Str]
#               <list of alternative flags, in preferred order>
our $CPP_STANDARD_FLAGS = {
  is_gcc => {
    'C++98' => [ "-std=c++98" ],
    'C++11' => [ "-std=c++11", "-std=c++0x" ],
    'C++14' => [ "-std=c++14", "-std=c++1y" ],
    'C++17' => [ "-std=c++17", "-std=c++1z" ],
  },
  is_clang => {
    'C++98' => [ "-std=c++98", ],
    'C++11' => [ "-std=c++11", ],
    'C++14' => [ "-std=c++14", "-std=c++1y" ],
    'C++17' => [ "-std=c++17", "-std=c++1z" ],
  },
  is_msvc => {
    # Newer MSVC set C++14 as minimum version.
    'C++98' => [ "" ],
    'C++11' => [ "" ],
    'C++14' => [ "-std:c++14" ],
    'C++17' => [ "-std:c++17" ],
  },
  is_sunstudio => {
    'C++98' => [ "" ],
    'C++11' => [ "-std=c++11", "-std=c++0x" ],
    'C++14' => [ "-std=c++14" ],
    # No mention of C++17 for Oracle Developer Studio 12.6.
  },
};

sub cpp_standard_flag {
  my ($self, $standard_name) = @_;

  $self->guess_compiler || die;
  my ($detected_compiler) = grep { $self->{$_} } keys %$CPP_STANDARD_FLAGS;

  die "Unknown standard '$standard_name' for compiler '$detected_compiler'"
    unless exists $CPP_STANDARD_FLAGS->{$detected_compiler}{$standard_name};

  my $test_flags = $CPP_STANDARD_FLAGS->{$detected_compiler}{$standard_name};

  for my $flag (@$test_flags) {
    return $flag if $self->_can_compile_code( <<EOF, [ $flag ] );
int main(){ return 0; }
EOF
  }

  die "Compiler '$detected_compiler' does not support any flags for standard '$standard_name'";
}

1;
